src/throttler/throttler.ts

```diff
@@ -91,20 +96,39 @@ export abstract class Throttler<T> {
     return p;
   }
 
-  public add(task: T): void {
+  /**
+   * Add the task to the throttler.
+   * When the task is completed, resolve will be called with handler's result.
+   * If this task fails after retries, reject will be called with the error.
+   */
+  public add(
+    task: T,
+    wait?: { resolve: (result: any) => void; reject: (err: Error) => void }
+  ): void {
```

Comment:
I'd vote for having `add` return promises for all cases, so we don't need two separate methods and separate code paths. If the caller does not need the result, they can feel free to discard it, just like calling any other function that returns a promise. A promise is not expensive to create and a promise without `.then` handlers will not cause any memory leaks. That's just my personal opinion though and you're welcome to keep it as it as.

Reply:
I agree with @yuchenshi 
@bkendall Thoughts on this.

Reply:
Why does the user of this function care about the `wait` argument? Basically, that's the thing that's exposing that it's implemented using Promises. If I `.add` something to a queue, I'd very much expect that to be synchronous. The thing it _does_ may be async, but adding to a queue should not require callbacks of any kind.

Not handling the promise _is_ a dangerous idea, as well. If it failed, you'd have an uncaught exception floating around, which will cause issues as well.

The public API for this should make sense for the user: they want to add a `task`. Does the _user_ care about the ability to get the results? I think that should be available to the user, but I _don't_ think this is the appropriate place to attach and get that information - it feels like `wait` is an implementation detail and that your first instinct is more right, that there should be a helper that wraps the task being added with whatever the Throttler needs, but does so synchronously.

Reply:
Sure, I will keep private `addHelper` to hide the wait argument.

Reply:
Did you mean having `add(task) -> void` (without `wait` argument) and `throttle(task) -> Promise`? If so, then totally agree with it.

However, I find the `throttle` method name confusing here because `add` also does throttling. It should have a name that signifies that it adds the task, and then **returns a promise that is resolved when the task finishes**. The bold part makes it different from `add`. Something like `addAndWaitForResult(task)` (but shorter) would be better IMO.

Reply:
`run` would be a OK candidate here.